### PR TITLE
fix: Make Image serialization background process

### PR DESCRIPTION
**Pull request by tssweeney**
This PR backgrounds & parallelizes Image file uploads. Calling 50 ops that make accept and return images goes from ~21s to ~2 seconds. Moreover, the upload time is nearly entirely at program exit.

(After  |    Before)
Screenshot 2024-09-10 at 19 48 59

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (tssweeney) [inline .vscode/settings.json:1]: This is intentionally checked in and helps vsocde know where to find our tests. Unrelated to images

File: weave/trace/async_job_queue.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import atexit
+import concurrent.futures
+import threading
+from concurrent.futures import Future
+from typing import Any, Callable, TypeVar
+
+T = TypeVar("T")
+
+MAX_WORKER_DEFAULT = 2**6
```
Comment (gtarpenning) [inline weave/trace/async_job_queue.py:9]:
```suggestion
MAX_WORKER_DEFAULT = 2**6  # 64
```

File: weave/trace/async_job_queue.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    Attributes:
+        executor (concurrent.futures.ThreadPoolExecutor): The executor used to run jobs.
+        _lock (threading.Lock): A lock to ensure thread-safe operations.
+        _is_shutdown (bool): A flag indicating whether the queue has been shut down.
+        _active_jobs (set): A set to keep track of active jobs and callbacks.
+        _max_workers (int): The maximum number of worker threads to use.
+
+    Args:
+        max_workers (int): The maximum number of worker threads to use. Defaults to 5.
```
Comment (gtarpenning) [inline weave/trace/async_job_queue.py:28]:
```suggestion
        max_workers (int): The maximum number of worker threads to use. Defaults to 64.
```

File: weave/trace/async_job_queue.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        If the queue has been shut down, it will be restarted automatically.
+
+        Args:
+            func: The function to be executed.
+            *args: Positional arguments to pass to the function.
+            **kwargs: Keyword arguments to pass to the function.
+
+        Returns:
+            A Future representing the execution of the job.
+        """
```
Comment (gtarpenning) [inline weave/trace/async_job_queue.py:63]: i feel like an example might be more useful here

File: weave/trace/weave_client.py [hunk 5]
```diff
@@ -408,7 +412,9 @@ def __init__(
         self.project = project
         self.server = server
         self._anonymous_ops: dict[str, Op] = {}
+        self.async_job_queue = AsyncJobQueue()
         self.ensure_project_exists = ensure_project_exists
+        atexit.register(self._cleanup)
```
Comment (gtarpenning) [inline weave/trace/weave_client.py:417]: should we put this on a dunder method? post_init? @andrewtruong would know where

File: weave/trace/async_job_queue.py (hunk 2 above)
Comment (gtarpenning) [reply weave/trace/async_job_queue.py:9]: too many?

**Review by gtarpenning — APPROVED** [review]

File: weave/trace/weave_client.py (hunk 5 above)
Comment (tssweeney) [reply weave/trace/weave_client.py:417]: Not sure what you are recommending here. We do call cleanup in __del__ as well.

File: weave/trace/async_job_queue.py (hunk 2 above)
Comment (tssweeney) [reply weave/trace/async_job_queue.py:9]: reduced to 8

File: weave/trace/async_job_queue.py (hunk 3 above)
Comment (tssweeney) [reply weave/trace/async_job_queue.py:28]: ty

File: weave/trace/async_job_queue.py (hunk 4 above)
Comment (tssweeney) [reply weave/trace/async_job_queue.py:63]: added

**Pull request merged**